src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java

```diff
@@ -1254,6 +1254,17 @@ private void initToken(CK_SLOT_INFO slotInfo) throws PKCS11Exception {
                 ("Token info for token in slot " + slotID + ":");
             System.out.println(token.tokenInfo);
         }
+
+        long[] supportedMechanisms = p11.C_GetMechanismList(slotID);
+        // start w/ the supported mechanisms, then remove the config disabled
+        // and known broken ones
+        Set<Long> enabledMechSet =
+                Arrays.stream(supportedMechanisms).boxed().collect
+                (Collectors.toCollection(HashSet::new));
+        if (config.getDisabledMechanisms() != null) {
+            enabledMechSet.removeAll(config.getDisabledMechanisms());
```

Comment:
Should `enabledMechSet` be further restricted to `Config::enabledMechanisms` (if set)? `Config::disabledMechanisms` looks like a fallback for when `Config::enabledMechanisms` is not set, according to `Config::isEnabled`. I'd keep the logic that makes `Config::enabledMechanisms` work in pair with `Config::disabledMechanisms` in a single place, as duplicating it into two different places may lead to misalignment.

Reply:
Yes, you have a good point. Let me think about it and see how to better handle this.